### PR TITLE
DAOS-13975 control: Store checker policies in MS only

### PR DESCRIPTION
**Pull request by kjacque**
This patch makes the control plane management service solely responsible for tracking checker policies. There is no longer any assumption that the daos_engine is maintaining any knowledge of policies.

- The MS stores both the set policy and the last policy used in a checker run. Both are persisted via raft.
- Add an option to "dmg check get-policy" to fetch the last policy used in a checker run.
- If no policies have been set, the MS sets all values to default.

Features: recovery control

Required-githooks: true

### Before requesting gatekeeper:

* [x] Two review approvals and any prior change requests have been resolved.
* [x] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by tanabarr — APPROVED** [review]

**Review by knard-intel — APPROVED** [review]

**Comment by Nasf-Fan** [comment]
I made some simple tests. There are something can be improved:

1. Before start checker for the first time, I run "dmg check get-policy -d", then it shows:

[fyong@boro-38 bin]$ ./dmg check get-policy -d
ERROR: dmg: rpc error: code = Unknown desc = no previous policies have been saved

That is not friendly output, instead, if checker has never been started, the output for "dmg check get-policy" and "dmg check get-policy -d" can be the same.

2. If I never run "dmg check set-policy", then the "dmg check get-policy" will show:

[fyong@boro-38 bin]$ ./dmg check get-policy
Checker flags: none

Inconsistency Class          Repair Action 
-------------------          ------------- 
POOL_LESS_SVC_WITH_QUORUM    DEFAULT       
POOL_LESS_SVC_WITHOUT_QUORUM DEFAULT       
POOL_MORE_SVC                DEFAULT       
POOL_NONEXIST_ON_MS          DEFAULT       
POOL_NONEXIST_ON_ENGINE      DEFAULT       
POOL_BAD_SVCL                DEFAULT       
POOL_BAD_LABEL               DEFAULT       
ENGINE_NONEXIST_IN_MAP       DEFAULT       
ENGINE_DOWN_IN_MAP           DEFAULT       
ENGINE_HAS_NO_STORAGE        DEFAULT       
CONT_NONEXIST_ON_PS          DEFAULT       
CONT_BAD_LABEL               DEFAULT       
DTX_CORRUPTED                DEFAULT       
DTX_ORPHAN                   DEFAULT       
CSUM_LOST                    DEFAULT       
CSUM_FAILURE                 DEFAULT       
OBJ_LOST_REP                 DEFAULT       
OBJ_LOST_EC_SHARD            DEFAULT       
OBJ_LOST_EC_DATA             DEFAULT       
OBJ_DATA_INCONSIST           DEFAULT

That is expected. But once I run "dmg check set-policy" for some inconsistency class, then only the specified policy is shown, all the others are disappeared:

[fyong@boro-38 bin]$ ./dmg check set-policy POOL_BAD_LABEL:CIA_TRUST_MS    
system checker policies updated
[fyong@boro-38 bin]$ ./dmg check get-policy   
Checker flags: none

Inconsistency Class Repair Action 
------------------- ------------- 
POOL_BAD_LABEL      TRUST_MS      

It is expected that the other non-specified inconsistency classes are also shown as 'DEFAULT' policy.

3. If I continuously set-policy for other different inconsistency class, then the former set one is overwritten. For example:

[fyong@boro-38 bin]$ ./dmg check set-policy CONT_BAD_LABEL:CIA_IGNORE      
system checker policies updated
[fyong@boro-38 bin]$ ./dmg check get-policy
Checker flags: none

Inconsistency Class Repair Action 
------------------- ------------- 
CONT_BAD_LABEL      IGNORE        


Please note that the former one "POOL_BAD_LABEL:CIA_TRUST_MS" is disappear. But "POOL_BAD_LABEL" and "CONT_BAD_LABEL" are two different inconsistency classes. Set one should not affect the other.

**Review by Nasf-Fan — CHANGES_REQUESTED** [review]
Please consider to improve the patch to resolve the issues mentioned in former comment. Thanks!

**Comment by kjacque** [comment]
> Please consider to improve the patch to resolve the issues mentioned in former comment. Thanks!

Good catches. I'll work on the fixes.

**Comment by kjacque** [comment]
Test failures:
- https://daosio.atlassian.net/browse/DAOS-14188
- https://daosio.atlassian.net/browse/DAOS-13998

File: src/control/server/mgmt_check.go [hunk 1]
```diff
@@ -283,6 +277,23 @@ func (svc *mgmtSvc) SystemCheckStart(ctx context.Context, req *mgmtpb.CheckStart
...
+	if err != nil {
+		return nil, err
+	}
+
+	// Allow the requested policies to override any policies stored in the policy map.
+	for _, pol := range policies {
+		pm[pol.InconsistCas] = pol
+	}
+	result := make([]*mgmtpb.CheckInconsistPolicy, 0, len(pm))
+	for _, pol := range pm {
```
Comment (tanabarr) [inline src/control/server/mgmt_check.go:291]: Rather than looping through, why not use copy() on pm.ToSlice()?

**Review by knard-intel — APPROVED** [review]

File: src/control/server/mgmt_check.go (hunk 1 above)
Comment (kjacque) [reply src/control/server/mgmt_check.go:291]: Ah, we could actually just use `pm.ToSlice()` without a copy. That's a good idea. I'll change it if there are any other change requests, if that's okay with you.

**Review by Nasf-Fan — APPROVED** [review]
The option for sub-command get-policy "Fetch the last policy used by the checker" is "-d | --latest". "--latest" is fine, but the "-d" is some misguided and easy to be mis-regarded as the abridged for "--default". But the real behavior is the opposite.

So if you have other chance/requirment to refresh the patch, it is suggested to use  more clear name for such options.

**Comment by kjacque** [comment]
> The option for sub-command get-policy "Fetch the last policy used by the checker" is "-d | --latest". "--latest" is fine, but the "-d" is some misguided and easy to be mis-regarded as the abridged for "--default". But the real behavior is the opposite.
> 
> So if you have other chance/requirment to refresh the patch, it is suggested to use more clear name for such options.

Good idea. "-l" is unfortunately taken by --host-list. Maybe --latest could use "-L" or not have a short form at all.

**Review by tanabarr — APPROVED** [review]

**Comment by kjacque** [comment]
Test failures are this pool consolidation test issue: https://daosio.atlassian.net/browse/DAOS-13998

**Review by Nasf-Fan — APPROVED** [review]

**Review by mjmac — APPROVED** [review]

**Pull request merged**